src/zarr/core/array.py

```diff
@@ -245,8 +240,34 @@ async def _create_v3(
         if not exists_ok:
             await ensure_no_existing_node(store_path, zarr_format=3)
 
-        shape = parse_shapelike(shape)
-        codecs = list(codecs) if codecs is not None else [BytesCodec()]
+        array_shape = parse_shapelike(shape)
+        shard_shape: tuple[int, ...] | None = None
+        chunk_shape: tuple[int, ...]
+
+        # because chunks is an optional typeddict with optional keys, it could be completely empty
+        # OR None, both of which result in chunks being inferred automatically
+        if chunks is not None and not (chunks == {}):
+            if isinstance(chunks, dict):
+                if "write_shape" in chunks:
+                    chunk_shape = chunks["write_shape"]
+                    if "read_shape" in chunks:
+                        # sharding is only enabled when read_shape and write_shape are specified
+                        # we do not special-case the condition when read_shape and write_shape are the same
+                        shard_shape = chunks["read_shape"]
+                elif "read_shape" in chunks:
+                    # if read_shape is present, but write_shape is absent, then
+                    # set the chunk_shape to read_shape, and keep shard_shape set to `None`
+                    chunk_shape = chunks["read_shape"]
+            else:
+                chunk_shape = parse_shapelike(chunks)
+        else:
+            # determine chunking parameters automatically
+            chunk_shape = _guess_chunks(array_shape, np.dtype(dtype).itemsize)
+
+        _codecs = tuple(codecs) if codecs is not None else (BytesCodec(),)
+
+        if shard_shape is not None:
+            _codecs = (ShardingCodec(chunk_shape=shard_shape, codecs=_codecs),)
```

Comment:
This effectively hard-codes sharding into the spec, something like a `sharded=True` flag that might have existed on the CHunkSpec object. How do you expect this to extend to variable chunking or other schemes that might be created in the future?

Reply:
This proposal can use whatever specification for variable length chunks we come up with, e.g. tuples of tuples of ints. You could specify variable length chunking with no sharding via something like `chunks = {'write_shape': ((10,5), (1,2,3)}`, and variable length chunking with sharding via something like `chunks = {'write_shape`: ((10,5), (1,2,3)), `read_shape`: (1,1)}`. The read shape would have to checked for consistency with all the unique chunk shapes in this case. We would of course need to widen the type of `ChunkSpec` for this to accept `tuple[tuple[int, ...]]` for the `write_shape` keys.

Reply:
> something like a sharded=True flag that might have existed on the CHunkSpec object.

If we had such a flag on the chunkspec object, then it would semantically collide with `read_shape`: `{'write_shape': (10,10), 'read_shape': (2,2), sharding: False}` would not be valid, because there's no way to have `read_shape` and `write_shape` differ without sharding. BTW when I say "sharding" i don't mean "the sharding codec", I mean the general concept of packing multiple subchunks into a single file. If a non-codec implementation of sharding emerges, then I would like to imagine that this API could wrap that.